tests/malloc.c

```diff
@@ -1,65 +1,100 @@
 #include <stdc.h>
 #include <malloc.h>
 #include <ktest.h>
+#include <thread.h>
+#include <sched.h>
 
-static int test_malloc() {
-  vm_page_t *page = pm_alloc(1);
+static int malloc_one_allocation(void) {
+  MALLOC_DEFINE(mp, "testing memory pool");
+  kmalloc_init(mp, 1, 1);
+  void *ptr = kmalloc(mp, 1234, M_NOWAIT);
+  assert(ptr != NULL);
+  kfree(mp, ptr);
+  return KTEST_SUCCESS;
+}
 
+static int malloc_invalid_values(void) {
   MALLOC_DEFINE(mp, "testing memory pool");
+  kmalloc_init(mp, 1, 1);
+  void *ptr = kmalloc(mp, PAGESIZE, M_NOWAIT);
+  assert(ptr == NULL);
+  ptr = kmalloc(mp, 0, M_NOWAIT);
+  assert(ptr == NULL);
+  return KTEST_SUCCESS;
+}
 
-  kmalloc_init(mp);
-  kmalloc_add_arena(mp, page->vaddr, PAGESIZE);
+static int malloc_multiple_allocations(void) {
+  MALLOC_DEFINE(mp, "testing memory pool");
+  kmalloc_init(mp, 1, 1);
+  const int n = 50;
+  void *ptrs[n];
+  for (int i = 0; i < n; i++)
+    assert(NULL != (ptrs[i] = kmalloc(mp, i + 1, M_NOWAIT)));
+  for (int i = 0; i < n; i++)
+    kfree(mp, ptrs[i]);
+  return KTEST_SUCCESS;
+}
 
-  void *ptr1 = kmalloc(mp, 15, 0);
+static int malloc_dynamic_pages_addition(void) {
+  MALLOC_DEFINE(mp, "testing memory pool");
+  kmalloc_init(mp, 1, 3);
+  void *ptr1 = kmalloc(mp, 4000, 0);
   assert(ptr1 != NULL);
-
-  void *ptr2 = kmalloc(mp, 23, 0);
-  assert(ptr2 != NULL && ptr2 > ptr1);
-
-  void *ptr3 = kmalloc(mp, 7, 0);
-  assert(ptr3 != NULL && ptr3 > ptr2);
-
-  void *ptr4 = kmalloc(mp, 2000, 0);
-  assert(ptr4 != NULL && ptr4 > ptr3);
-
-  void *ptr5 = kmalloc(mp, 1000, 0);
-  assert(ptr5 != NULL);
-
+  void *ptr2 = kmalloc(mp, 4000, 0);
+  assert(ptr2 != NULL);
+  void *ptr3 = kmalloc(mp, 4000, 0);
+  assert(ptr3 != NULL);
   kfree(mp, ptr1);
   kfree(mp, ptr2);
-  kmalloc_dump(mp);
   kfree(mp, ptr3);
-  kfree(mp, ptr5);
+  return KTEST_SUCCESS;
+}
 
-  void *ptr6 = kmalloc(mp, 2000, M_NOWAIT);
-  assert(ptr6 == NULL);
+#define ALLOCATIONS_PER_THREAD 1000
+#define THREADS_NUMBER 10
 
-  pm_free(page);
+static void threads_function(void *arg) {
+  for (int i = 0; i < ALLOCATIONS_PER_THREAD; i++) {
+    void *ptr = kmalloc(arg, 1 + (i * (int)arg) % 32, 0);
+    assert(ptr != NULL);
```

Comment:
Yes, but size is != 0, there is no `M_NOWAIT` and I can't find panic message "memory exhausted in ".